### ISSUE TITLE
Allow onShippingChange reject to supply custom message

### ISSUE THREAD

**Issue by cabintech** [paypal/paypal-checkout-components#1140]
Currently if we have to reject a customer address change in onShippingChange(), the PayPal dialog show a generic, and possibly misleading message "My Company doesn't ship to this location. Please use a different address."

There are many reasons an address might be rejected, not all of them as simple as that message suggests. For example, the customer has chosen a shipping service which is not valid for that address, but there are other shipping services we offer which can deliver to that address. They read that message, assume we cannot ship to them, abandon the cart, and we lose the sale. Not good.

We need some control over the message displayed. Since we are forced to supply an Error object on the reject(), why not display the message text from that? I would even like the ability to halt the flow at this point and return to our shopping cart, but not sure how to make that work with reject() and resolve() as the only options.

**Comment by springknees** [comment]
I agree! My very simple implementation requires the purchaser to enter their shipping destination on our purchase page. I then compute the shipping cost and present the Smart Checkout Button together with the computed shipping cost (to the selected destination). A problem arises because the purchaser can then select another destination on the Paypal portal (either accidentally or deliberately), which can then enable them to pay a reduced rate for shipping - not good!

I was hoping to that the Paypal Javascript SDK onShippingChange() would enable me to prevent this scenario - which to it does to some degree.

As @cabintech says the message "My Company doesn't ship to this location. Please use a different address." doesn't say what I'd like to say, which is something like "Your shipping destination does not match the destination you selected!"

Also, as @cabintech suggests, I would like the ability to halt the flow at this point and return to our shopping cart so that the purchaser can select the correct shipping destination.

**Comment by bluepnume** [comment]
@mcntsh can you follow up on if we have plans to support a custom message?

**Comment by mcntsh** [comment]
We're more inclined to make the message either more "generic" to account for a wider range of use-cases, or to give the merchant the option to pass an enum to the `actions.reject()` method to change the context of the error.

@cabintech, it seems like there's an incompatibility between the address the buyer provided before checkout and what they selected within checkout. Traditionally, we'd solve this by having you (the merchant) send the buyer's address within the Order Create call so that it can no longer be changed within the checkout flow. 

```
return actions.order.create({
    application_context: {
        shipping_preference: 'SET_PROVIDED_ADDRESS'  
    },
    purchase_units: [{
        shipping: {
            address: {
                address_line_1: '555 Example Lane',
                admin_area_2: 'San Jose',
                admin_area_1: 'CA',
                postal_code: '95111',
                country_code: 'US'
            }
        },
        amount: {
            value: '13.37'
        }
    }]
});
``` 

This might be a good solution for @springknees, since they have the expectation where the buyer should not change that address, but if you don't care if the address is changed as long as whatever they picked is still eligible for the original shipping service, this would indeed be a good use of the `onShippingChange` callback.

I'll take this request back to our content team to see what they come up with.

**Comment by cabintech** [comment]
In our case, we know only enough of the address (country, zip) to have the user select a shipping vendor and shipping service, from which we calculate shipping costs. We really don't want to collect full addresses and we certainly don't want to restrict what address they can use once they get into the payment process. So the onShippingChange callback is ideal... except that there are a lot of possible failure modes and we need more control on communicating with the customer after an address change.

IMHO you will never be able to formulate a message that makes sense for all the possible reasons a merchant might reject a shipping change. There are merchant specific complexities and subtleties you cannot anticipate. Why the reluctance to let the merchant supply the message? If they don't supply one, then go ahead with some kind of generic message.

**Comment by mcntsh** [comment]
> Why the reluctance to let the merchant supply the message?

Security, mainly. Allowing for the error message to be freely customizable has potential for abuse. Imagine a rogue merchant sends through a message saying "Contact the PayPal support line at XXX-XXX-XXXX". It's displayed within the PayPal flow and it looks like an official error message. Imagine the buyer contacts that number and is asked for their password by someone pretending to be PayPal.

I don't think we will be able to formulate a message that makes sense for *all* scenarios within this callback, which is why I'd like to defer certain use-cases to other features as they make sense. Just because something can be solved with the callback doesn't mean it should. 

As it relates to your problem, we can simplify the error to there being some kind of incompatibility with the buyer's shipping address and what the merchant expects. That's a generic enough scenario that we can write a message for. Or perhaps if we need to get more granular we can, and provide merchants with a choice between different rejection errors.

**Comment by cabintech** [comment]
Oh, yup, I get it. Did not think of that... here I was assuming that merchants are trusted! Ha!

**Comment by springknees** [comment]
Thanks @mcntsh. I like this-

> we can simplify the error to there being some kind of incompatibility between what the buyer's shipping address is and what the merchant expects it to be. That's a generic enough scenario that we can write a message for. Or perhaps if we need to get more granular we can, and provide merchants with a choice between different rejection errors.

@cabintech originally requested the ability to halt the flow on `actions.reject() ` - if there was also the option to pass an enum to the `actions.reject()` method (as you suggest), that would be ideal.

**Comment by evamedia** [comment]
How about a choice of messages to choose, my client will ship anywhere, but can't standardise for certain countries, so would like to return "Please contact us  to purchase this item with the correct shipping for your country."

**Comment by bluepnume** [comment]
Closing this for now. We can reconsider this request in future, but for now we don't have plans to change the default behavior. Thanks.
